wire-ios-data-model/Source/MLS/MLSGroupVerification.swift

```diff
@@ -0,0 +1,108 @@
+// Copyright (C) 2024 Wire Swiss GmbH
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+
+// sourcery: AutoMockable
+public protocol MLSGroupVerificationProtocol {
+
+    func startObserving()
+    func updateConversation(by groupID: MLSGroupID) async
+    func updateConversation(_ conversation: ZMConversation, with groupID: MLSGroupID) async
+    func updateAllConversations() async
+
+}
+
+public final class MLSGroupVerification: MLSGroupVerificationProtocol {
+
+    // MARK: - Properties
+
+    private let updateUseCase: any UpdateMLSGroupVerificationStatusUseCaseProtocol
+
+    private let mlsService: MLSServiceInterface
+    private let syncContext: NSManagedObjectContext
+
+    private var observationTask: Task<Void, Never>?
+
+    // MARK: - Initialize
+
+    public init(
+        updateUseCase: any UpdateMLSGroupVerificationStatusUseCaseProtocol,
+        mlsService: any MLSServiceInterface,
+        syncContext: NSManagedObjectContext
+    ) {
+        self.updateUseCase = updateUseCase
+        self.mlsService = mlsService
+        self.syncContext = syncContext
+    }
+
+    deinit {
+        observationTask?.cancel()
+    }
+
+    // MARK: Observing
+
+    public func startObserving() {
+        observationTask = .detached { [weak self] in
+            guard let asyncStream = self?.mlsService.epochChanges() else { return}
+
+            for await groupID in asyncStream {
+                if Task.isCancelled { return }
+                await self?.updateConversation(by: groupID)
+            }
+        }
+    }
+
+    // MARK: Update Conversation
+
+    public func updateConversation(by groupID: MLSGroupID) async {
+        guard let conversation = await syncContext.perform({
+            ZMConversation.fetch(with: groupID, in: self.syncContext)
+        }) else {
+            return WireLogger.e2ei.warn("failed to fetch the conversation by mlsGroupID \(groupID)")
+        }
+
+        await updateConversation(conversation, with: groupID)
+    }
+
+    public func updateConversation(_ conversation: ZMConversation, with groupID: MLSGroupID) async {
+        do {
+            try await updateUseCase.invoke(for: conversation, groupID: groupID)
+        } catch {
+            WireLogger.e2ei.warn("failed to update MLS group: \(groupID.safeForLoggingDescription) verification status: \(String(describing: error))")
```

Comment:
I would keep "failed to update conversation's verification status", because it's more accurate. We only update the verification status. 
I would also rename `updateUseCase` to something like `updateVerificationStatus`